projects/asn1-proto/asn1_proto_to_der.cc

```diff
@@ -0,0 +1,153 @@
+#include "asn1_proto_to_der.h"
+
+namespace asn1_proto {
+
+uint8_t ASN1ProtoToDER::GetNumBytes(const size_t num, const size_t base) {
+  uint8_t base_bits = log2(base);
+  for (uint8_t num_bits = sizeof(num) * 8; num_bits != 0; --num_bits) {
+    if ((num >> num_bits) & 0x01) {
+      return ceil(num_bits / base_bits);
+    }
+  }
```

Comment:
```suggestion
  for (uint8_t num_bits = sizeof(num) * 8; num_bits > base_bits; num_bits -= base_bits) {
    if ((num >> num_bits)) {
      return ceil(num_bits / base_bits);
    }
  }
```

You don't need the mask here, because any bit being 1 counts.

The further optimization is to check `base_bits` worth of bits at a time (i.e. either 7 or 8 bits). This is in effect 'loop unrolling'.

For example, for a base-128 number, it's going to do:
base_bits = 7
num_bits = 64
num_bits = 57
num_bits = 50
num_bits = 43
num_bits = 36
num_bits = 29
num_bits = 22
num_bits = 15
num_bits = 8
num_bits = 1  // terminate the loop

It doesn't matter if the last bit is set or not because we know both 0 and 1 require 1 byte to represent.

Although you probably want to spot-check my math here :)

Reply:
Ah, very nice, change accepted!